numba/cuda/compiler.py

```diff
@@ -943,6 +787,66 @@ def __init__(self, py_func, sigs, targetoptions):
             self.compile(sigs[0])
             self._can_compile = False
 
+        dispatcher = self
+
+        class function_template(AbstractTemplate):
+            key = dispatcher
+
+            def generic(self, args, kws):
+                assert not kws
+                return dispatcher.compile_device(args).signature
+
+            def get_template_info(cls):
+                basepath = os.path.dirname(os.path.dirname(numba.__file__))
+                code, firstlineno = inspect.getsourcelines(py_func)
+                path = inspect.getsourcefile(py_func)
+                sig = str(utils.pysignature(py_func))
+                info = {
+                    'kind': "overload",
+                    'name': getattr(cls.key, '__name__', "unknown"),
+                    'sig': sig,
+                    'filename': utils.safe_relpath(path, start=basepath),
+                    'lines': (firstlineno, firstlineno + len(code) - 1),
+                    'docstring': py_func.__doc__
+                }
+                return info
+
+        typingctx = cuda_target.typing_context
+        typingctx.insert_user_function(dispatcher, function_template)
+
+        name = getattr(py_func, '__name__', 'unknown')
+        self.__name__ = f"{name} <CUDA device function>".format(name)
+
+    def get_call_template(self, args, kws):
+        # Copied and simplified from _DispatcherBase.get_call_template.
+        """
+        Get a typing.ConcreteTemplate for this dispatcher and the given
+        *args* and *kws* types.  This allows to resolve the return type.
+
+        A (template, pysig, args, kws) tuple is returned.
+        """
+        # Ensure an overload is available
+        self.compile_device(tuple(args))
+
+        # Create function type for typing
+        func_name = self.py_func.__name__
+        name = "CallTemplate({0})".format(func_name)
+
+        # The `key` isn't really used except for diagnosis here,
+        # so avoid keeping a reference to `cfunc`.
+        call_template = typing.make_concrete_template(
+            name, key=func_name, signatures=self.nopython_signatures)
+        pysig = utils.pysignature(self.py_func)
+
+        return call_template, pysig, args, kws
+
+    def get_overload(self, sig):
+        # NOTE: This dispatcher seems to be used as the key for the dict of
+        # implementations elsewhere in Numba, so we return this dispatcher
+        # instead of a compiled entry point as in
+        # _DispatcherBase.get_overload().
+        return self
```

Comment:
This should be understood (or changed) before this PR is considered done:
* Why is `self` acceptable for `get_overload`?
* Why does the dispatcher and not the overload get used as the key for dicts in the CUDA target?